### PR TITLE
Story: API call Exchange Service (GET /catalogue)

### PR DESCRIPTION
**Pull request by tomling**
- GetCatalogue in ProductEditionService returns latest baseline, releasable versions for all products.
- GetCatalogue on CatalogueApiController implemented to call service methods.
- Unit tests around service and controller methods.
- Refactored existing logic to use async

### CODE REVIEW

**Review by bonpoole — APPROVED** [review]

File: src/UKHO.SalesCatalogueStub.Api/Services/ProductEditionService.cs [hunk 1]
```diff
@@ -339,6 +418,28 @@ private static Cancellation GetCancellation(int latestUpdateNumber)
...
+        private static Enc DeserializePidTombstone(string xmlString)
+        {
+            var serializer =
+                new XmlSerializer(typeof(Enc));
+
+            var stringReader = new StringReader(xmlString);
+
+            var enc = (Enc)serializer.Deserialize(stringReader);
+
+            return enc;
```
Comment (UKHOGreg) [inline src/UKHO.SalesCatalogueStub.Api/Services/ProductEditionService.cs:441]: Are any `using` statements required here?

File: src/UKHO.SalesCatalogueStub.Api/Services/ProductEditionService.cs [hunk 2]
```diff
@@ -210,6 +215,89 @@ requestUpdateNumber > (productDbMatch.UpdateNumber ?? 0)))
             return (matchedProducts, GetProductVersionResponseEnum.UpdatesFound);
         }
 
+        public bool CheckIfCatalogueModified(DateTime? ifModifiedSince, out DateTime? dateEntered)
+        {
+            var latestDateEntered = _dbContext.LoaderStatus.OrderByDescending(a => a.DateEntered).First(a => a.AreaName == AreaNameEnum.Main).DateEntered;
+
+            dateEntered = latestDateEntered;
+            return ifModifiedSince == null || ifModifiedSince < latestDateEntered;
+        }
```
Comment (UKHOGreg) [inline src/UKHO.SalesCatalogueStub.Api/Services/ProductEditionService.cs:224]: Can we make this `async`?

File: src/UKHO.SalesCatalogueStub.Api/Controllers/CatalogueApiController.cs [hunk 3]
```diff
@@ -40,38 +53,20 @@ public class CatalogueApiController : ControllerBase
...
-
-            //TODO: Uncomment the next line to return response 401 or use other options such as return this.NotFound(), return this.BadRequest(..), ...
-            // return StatusCode(401);
-
-            //TODO: Uncomment the next line to return response 403 or use other options such as return this.NotFound(), return this.BadRequest(..), ...
-            // return StatusCode(403);
 
-            //TODO: Uncomment the next line to return response 404 or use other options such as return this.NotFound(), return this.BadRequest(..), ...
-            // return StatusCode(404, default(DefaultErrorResponse));
+            var catalogueModified = _productEditionService.CheckIfCatalogueModified(ifModifiedSince, out var dateEntered);
```
Comment (UKHOGreg) [inline src/UKHO.SalesCatalogueStub.Api/Controllers/CatalogueApiController.cs:59]: Can we make this `async`?

File: src/UKHO.SalesCatalogueStub.Api.Tests/ProductEditionServiceGetCatalogueTests.cs [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+            _service = new ProductEditionService(new SalesCatalogueStubDbContext(dbContextOptions), A.Fake<ILogger<ProductEditionService>>());
+
+            CreateCatalogue();
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            _dbContext.Database.EnsureDeleted();
+        }
```
Comment (UKHOGreg) [inline src/UKHO.SalesCatalogueStub.Api.Tests/ProductEditionServiceGetCatalogueTests.cs:386]: Can we move these to the top of the class, just underneath the fields please?

**Review by UKHOGreg — CHANGES_REQUESTED** [review]
Really good, just a couple things!

File: src/UKHO.SalesCatalogueStub.Api/Services/ProductEditionService.cs (hunk 2 above)
Comment (tomling) [reply src/UKHO.SalesCatalogueStub.Api/Services/ProductEditionService.cs]: Resolved, thanks!

File: src/UKHO.SalesCatalogueStub.Api/Controllers/CatalogueApiController.cs (hunk 3 above)
Comment (tomling) [reply src/UKHO.SalesCatalogueStub.Api/Controllers/CatalogueApiController.cs]: Resolved, thanks!

File: src/UKHO.SalesCatalogueStub.Api.Tests/ProductEditionServiceGetCatalogueTests.cs (hunk 4 above)
Comment (tomling) [reply src/UKHO.SalesCatalogueStub.Api.Tests/ProductEditionServiceGetCatalogueTests.cs]: Resolved, thanks!

File: src/UKHO.SalesCatalogueStub.Api/Services/ProductEditionService.cs (hunk 1 above)
Comment (tomling) [reply src/UKHO.SalesCatalogueStub.Api/Services/ProductEditionService.cs]: Resolved, thanks!

**Review by UKHOGreg — APPROVED** [review]

**Pull request merged**